lib/task_listener.js

```diff
@@ -314,11 +319,102 @@ export default class TaskListener extends EventEmitter {
+   * Coalesce the task in the given claim, claiming any additional tasks
+   * directly from the TC queue service, then calling runTask with the
+   * resulting task set.
+   */
+  async coalesceClaim(claim) {
+    let task = claim.task;
+    let taskId = claim.status.taskId;
+
+    try {
+      // if the task is not coalescible, then the taskset is just the one claim
+      if (!task.payload.coalescer || !task.payload.coalescer.url) {
+        return [claim];
+      }
+
+      let routes = task.routes;
+      routes = routes.filter(r => r.startsWith(COALESCER_ROUTE_PREFIX));
+      if  (routes.length != 1) {
+        return [claim];
+      }
+
+      let coalescingKey = routes[0].substr(COALESCER_ROUTE_PREFIX.length);
+      let coalescerUrl = task.payload.coalescer.url + coalescingKey;
+
+      let tasks = await this.fetchCoalescerTasks(coalescerUrl);
+      if (tasks.length == 0) {
+        return [claim];
+      }
+
+      if (!tasks.some(tid => tid == taskId)) {
+        this.runtime.log('bad coalescer response', {
+          taskId: taskId,
+          runId: claim.runId,
+          response: tasks,
+          message: "taskIds do not include " + taskId,
+        });
+        return [claim]
+      }
+
+      // claim runId 0 for each of those tasks; we can consider adding support
+      // for other runIds later.
+      var claims = await Promise.all(tasks.map(async tid => {
+        // for the existing claim, just return it
+        if (tid == taskId) {
+          return claim;
+        }
+
+        try {
+          return await this.runtime.queue.claimTask(tid, 0, {
+            workerId: this.runtime.workerId,
+            workerGroup: this.runtime.workerGroup,
+          });
+        } catch(e) {
+          this.runtime.log("secondary claim failure", {
+            taskId: tid,
+            runId: 0,
+            message: e.toString(),
+            stack: e.stack,
+            err: e
+          });
+          return;
+        }
+      }));
+      return claims.filter(cl => cl);
+    } catch (e) {
+      this.runtime.log('coalescing error', {
+        taskId: claim.status.taskId,
+        runId: claim.runId,
+        message: e.toString(),
+        stack: e.stack,
+        err: e
+      });
+
+      // fail quietly by just returning the primary claim
+      return [claim];
+    }
+  }
+
+  async fetchCoalescerTasks(url) {
+    let jsonData = await request.get(url).timeout(this.coalescerTimeout).buffer().end();
+    if (!jsonData.ok || !jsonData.text) {
+      throw new Error("Failure fetching from coalescer URL " + url);
+    }
+    return JSON.parse(jsonData.text);
+  }
+
+  /**
   * Run task that has been claimed.
   */
-  async runTask(claim) {
+  async runTaskset(claims) {
     let runningState;
     try {
+      // the claim we're actually going to execute (the primary claim) is the
+      // first one; the rest will be periodically reclaimed and then resolved,
+      // but will not actually execute.
+      let claim = claims[0];
```

Comment:
scoping 'claim' here means that it's not available if an error is caught lower down.